streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```
@@ -667,6 +672,25 @@ private void addTasksToStateUpdater() {
         }
     }
 
+    public void tryHandleExceptionsFromStateUpdater() {
+        if (stateUpdater != null) {
+            final Map<TaskId, RuntimeException> taskExceptions = new LinkedHashMap<>();
+
+            for (final StateUpdater.ExceptionAndTasks exceptionAndTasks : stateUpdater.drainExceptionsAndFailedTasks()) {
+                final RuntimeException exception = exceptionAndTasks.exception();
+                final Set<Task> failedTasks = exceptionAndTasks.getTasks();
+
+                for (final Task failedTask : failedTasks) {
+                    // need to add task back to the bookkeeping to be handled by the stream thread
+                    tasks.addTask(failedTask);
+                    taskExceptions.put(failedTask.id(), exception);
+                }
+            }
+
+            maybeThrowTaskExceptions(taskExceptions);
```

Here we basically transform a mapping from exception -> set of tasks to a mapping task ID -> exception and then in `maybeThrowTaskExceptions()` we again transform the mapping task ID -> exception to a mapping exception -> set of tasks. 

Can we not make this a bit more straightforward by skipping the transformation to task ID -> exception?
 
Is it really necessary to aggregate all `TaskCorruptedException` into a new one? I think by re-throwing a brand new `TaskCorruptedException` we lose the information where the actual exception happened in the stacktrace which might make debugging harder. 

I see that with aggregating the exception we can revive all corrupted tasks all at once, but I am not sure if it is worth losing debugging information. One option could be to not drain the exceptions from the state updater, but handle one after the other. Another option could be to collect the `TaskCorruptedException` in the new exception and log all of them when the exception is caught.

---

The motivation for `maybeThrowTaskExceptions` is that 1) if there's exceptions that would cause "all is lost", then we can directly throw that one since we are going to close-dirty all tasks; 2) otherwise, then there's only task-corrupted, then we should handle them at once. If we do not do 2) then we need to bookkeep the list of "yet handled corrupted exception" to let the thread throw and handle one by one, which would be quite complicated.. I agree that doing two transformations here is a bit awkward, we can instead just do the quoted logic 1/2 here inside the function directly (from state updater, we should either see fatal streams exception or task corrupted only), I'm doing this just for paranoid code-deduping.

On the lose log information, yeah I totally agree with you. After thinking about that a bit I think we can just log the stack trace as we group the task corrupted exception, wdyt? BTW in the future we would finer-categorize different causes of task-corrupted as different exceptions so this logic is likely going to be temporary only.